### ISSUE TITLE
NeoVim integration

### ISSUE THREAD

**Issue by UberLambda** [zedapp/zed#511]
Would integrating Zed with [NeoVim's plugin API](https://github.com/neovim/neovim/wiki/Plugin-UI-architecture) be feasible without having to edit Zed's core too much? ACE's vi support is unfortunely a bit lacking, but using NeoVim a Zed applications set to use NeoVim as the editor backend would behave pretty much like Vim, while being much more flexible about plugins and the GUI.

**Comment by UberLambda** [comment]
Sorry for the double title editing, I couldn't make up my mind about wheter putting the question mark at the end or not to :P

**Comment by zefhemel** [comment]
Isn't this a plug-in API to extend neovim itself, rather than to use neovim as an editing component in another editor like Zed? Based on a quick look at the linked page I don't see how integration with Zed would really work. Do you have a more clear idea of this?

**Comment by UberLambda** [comment]
Yes, Zed would actually behave as a NeoVim plugin; however, a NVim plugin has a lot of freedom, and GUIs in fact can embed it. A NVim "plugin" can intercept and dispatch any vim events that it may want to handle - even custom ones, execute arbitrary vimscript code and generally it can modify the editor's state easilly. This means that Zed would still have complete control over its editor, and it would simply need to send events to NVim to notify it to write the file to disk, or open another file, etc.

**Comment by zefhemel** [comment]
So here's what I don't understand -- presumably this would run Neovim in some sort of headless mode, right? You woulnd't see the neovim UI, but use to interact with it? So basically every key press is then forwarded to the neovim process? Does Neovim have the ability to send back the current screen state? I don't see that anywhere.

Either way, this would be a pretty massive undertaking just to get better vim support :wink:

**Comment by UberLambda** [comment]
You can get each of neovim's buffers separately through the Api and then render it as you like. But yeah, if the core wasn't meant to support multiple backends, there would be alot to be rewritten...
I've posted this since someone mentioned easymotion support - that would be native through NeoVim - and because I think Zed could be a very nice, zero-distractions UI on top of Vim

**Comment by zefhemel** [comment]
Isn't vim pretty zero-distraction already?

**Comment by UberLambda** [comment]
Yes, but the problem is that it's interface is text-based. While it's
useful for SSH, the lack of a real GUI prevents the creation of some
plugins that require one (like the markdown preview inside the editor)
without resorting to external programs, while Zed+vim would allow it.
Il 27/feb/2015 16:26 "Zef Hemel" <notifications@github.com> ha scritto:
